dvc/ignore.py

```diff
@@ -75,18 +81,29 @@ class DvcIgnoreFilter(object):
     def __init__(self, root_dir, tree):
         self.tree = tree
         self.ignores = {DvcIgnoreDirs([".git", ".hg", ".dvc"])}
-        self._update(root_dir)
-        for root, dirs, _ in self.tree.walk(root_dir, dvcignore=self):
-            for d in dirs:
-                self._update(os.path.join(root, d))
 
     def _update(self, dirname):
         ignore_file_path = os.path.join(dirname, DvcIgnore.DVCIGNORE_FILE)
         if self.tree.exists(ignore_file_path):
             self.ignores.add(DvcIgnorePatterns(ignore_file_path, self.tree))
 
+    def _update_parents(self, path):
+        if not os.path.isabs(path):
+            path = os.path.join(self.tree.tree_root, path)
+        while path != self.tree.tree_root:
+            self._update(path)
+            prev_path = path
+            path = os.path.dirname(path)
+            if prev_path == path:
+                break
+        self._update(path)
+
     def __call__(self, root, dirs, files):
+        self._update_parents(root)
```

Comment:
In previous implementation we used to load all `.dvcignores` on `DvcIgnoreFilter` initialization. To avoid recursive `tree.walk -> DvcIgnoreFilter -> tree.walk` it needs to be moved out of `DvcIgnoreFilter.__init__`. 
Pro of current approach:
- we are sure that for each call we get up-to-date state of dvcignores in repo
Con:
- for each dir we will be visiting during `walk` we reload `dvcignores` from the `tree_root` up to currently visited `dir` which is  more resource-hungry than previous approach

Reply:
@pared maybe we could do that on tree initialization?